marimo/_save/cache.py

```diff
@@ -6,36 +6,102 @@
 from typing import TYPE_CHECKING, Any, Literal
 
+from marimo._plugins.ui._core.ui_element import UIElement
 from marimo._runtime.context import get_context
+from marimo._runtime.state import SetFunctor
 
 if TYPE_CHECKING:
+    from types import FrameType
+
     from marimo._ast.visitor import Name
 
-CacheType = Literal["ContentAddressed", "ExecutionPath", "Unknown"]
+CacheType = Literal[
+    "ContextExecutionPath",
+    "ContentAddressed",
+    "ExecutionPath",
+    "Pure",
+    "Unknown",
+]
 # Easy visual identification of cache type.
 CACHE_PREFIX: dict[CacheType, str] = {
+    "ContextExecutionPath": "X_",
     "ContentAddressed": "C_",
     "ExecutionPath": "E_",
+    "Pure": "P_",
     "Unknown": "U_",
 }
 
 ValidCacheSha = namedtuple("ValidCacheSha", ("sha", "cache_type"))
 
 
+# BaseException because "raise _ as e" is utilized.
+class CacheException(BaseException):
+    pass
+
+
 @dataclass
 class Cache:
     defs: dict[Name, Any]
     hash: str
+    stateful_refs: set[str]
     cache_type: CacheType
     hit: bool
 
+    def restore(self, frame: FrameType) -> None:
+        for lookup, var in self.contextual_defs():
+            frame.f_locals[var] = self.defs[lookup]
+
+        defs = {**globals(), **frame.f_locals}
+        for ref in self.stateful_refs:
+            if ref not in defs:
+                raise CacheException(
+                    "Failure while restoring cached values. "
+                    "Cache expected a reference to a "
+                    f"variable that is not present ({ref})."
+                )
+            value = defs[ref]
+            if isinstance(value, SetFunctor):
+                value(self.defs[ref])
+            elif isinstance(value, UIElement):
+                value._update(self.defs[ref])
+            else:
+                raise CacheException(
+                    "Failure while restoring cached values. "
+                    "Unexpected stateful reference type "
+                    f"({type(ref)}:{ref})."
+                )
+
+    def load(self, frame: FrameType) -> None:
+        for lookup, var in self.contextual_defs():
+            self.defs[lookup] = frame.f_locals[var]
+
+        defs = {**globals(), **frame.f_locals}
+        for ref in self.stateful_refs:
+            if ref not in defs:
+                raise CacheException(
+                    "Failure while saving cached values. "
+                    "Cache expected a reference to a "
+                    f"variable that is not present ({ref})."
+                )
+            value = defs[ref]
+            if isinstance(value, SetFunctor):
+                self.defs[ref] = value._state()
+            elif isinstance(value, UIElement):
+                self.defs[ref] = value.value
+            else:
+                raise CacheException(
+                    "Failure while saving cached values. "
+                    "Unexpected stateful reference type "
+                    f"({type(ref)}:{ref})."
+                )
```

Comment:
Are `restore` and `load` the same function? Should one be deleted?

Reply:
Will put in doc-strings too, but I think `update` might be a better name for cache

`update`: globals -> backfill object to be saved
`restore`: saved object -> globals

lmk if you have better name ideas!

Reply:
Ah I see when loads from defs -> frame, the other loads from frame -> defs, I missed that. Yea `update` seems like a good name.